fix: use ANSI colors codes instead of RGB

DanielleMaywood opened the pull request:
Fixes https://github.com/coder/coder/issues/13211

Instead of using RGB color codes, we instead use ANSI color codes. This means that we respect a user's terminal theme. Not all of the new colors match the old ones but this is a limitation with using 4-bit ANSI colors.

This change also allows users to pass `--no-color` to the CLI to completely disable theming. It is worth noting that whilst we do define the `serpent.Option`, we manually check `os.Args` for the flag as there doesn't appear to be a way to have a global middleware with `serpent` (whereas `spf13/cobra` does have this).

This change also removes `> ` from Prompt, and replaces it with `  ` as on `--no-color` there is no difference.

Before:
<img width="385" alt="Screenshot 2024-09-13 at 15 08 26" src="https://github.com/user-attachments/assets/832bef96-613b-4627-899d-2a41471db52a">

After:
<img width="559" alt="Screenshot 2024-09-13 at 15 09 15" src="https://github.com/user-attachments/assets/f4b0019c-0e2b-472c-ba0e-f11c716a540f">

Before:
<img width="377" alt="Screenshot 2024-09-13 at 15 08 34" src="https://github.com/user-attachments/assets/65b43199-a3ba-42cb-bfa1-c99ba9fa8fa6">

After:
<img width="563" alt="Screenshot 2024-09-13 at 15 09 30" src="https://github.com/user-attachments/assets/bf8d4ca5-a652-4555-b389-6c650914a132">

Before:
<img width="387" alt="Screenshot 2024-09-13 at 15 08 45" src="https://github.com/user-attachments/assets/cd1f4e6f-d6f5-48bd-8235-98807f0d767c">

After:
<img width="567" alt="Screenshot 2024-09-13 at 15 09 39" src="https://github.com/user-attachments/assets/24037e83-758f-49f1-a187-e31158e59043">

Before:
<img width="380" alt="Screenshot 2024-09-13 at 15 08 54" src="https://github.com/user-attachments/assets/541c495e-9d95-47ad-a5f8-0df5d681a7ea">

After:
<img width="555" alt="Screenshot 2024-09-13 at 15 09 50" src="https://github.com/user-attachments/assets/cc0733f4-dd1d-4202-a215-db01873cf640">

~~With `--no-color`~~ Now requires `NO_COLOR` environment variable
<img width="663" alt="Screenshot 2024-09-13 at 15 10 09" src="https://github.com/user-attachments/assets/837ab6b1-7f6a-48c3-8f42-574575af1458">

DanielleMaywood committed:
chore: add command for showing colors

DanielleMaywood committed:
fix: use ANSI color codes instead of RGB

DanielleMaywood committed:
feat: add '--no-color' flag

DanielleMaywood committed:
fix: revert colors

DanielleMaywood committed:
chore: change colors

DanielleMaywood committed:
fix: update golden files

DanielleMaywood committed:
fix: replace blue with brightBlue

DanielleMaywood committed:
fix: drop '> ' for unfocused prompts

DanielleMaywood committed:
fix: run 'make fmt'

johnstcn commented on cli/root.go:
```diff
@@ -461,6 +461,12 @@ func (r *RootCmd) Command(subcommands []*serpent.Command) (*serpent.Command, err
 			Group:       globalGroup,
 		},
+		{
+			Flag:        cliui.NoColorFlag,
```
Wonder if it would make sense to also allow setting it as an env (`CODER_NO_COLOR`)? Not blocking.

DanielleMaywood replied:
Yes that should be a good idea. There is a 'standard' around using `NO_COLOR` for this https://no-color.org. We could do both `NO_COLOR` and `CODER_NO_COLOR`?

johnstcn replied:
Sounds good to me!

johnstcn commented on cli/cliui/cliui.go:
```diff
@@ -37,22 +43,33 @@ var (
-	Fuchsia = Color("#EE6FF8")
-	Yellow  = Color("#ECFD65")
-	Blue    = Color("#5000ff")
+	red           = Color("1")
+	green         = Color("2")
+	yellow        = Color("3")
+	magenta       = Color("5")
+	white         = Color("7")
+	brightBlue    = Color("12")
+	brightMagenta = Color("13")
```
nit: indicate these are ANSI color codes

johnstcn commented on cli/cliui/cliui.go:
```diff
@@ -109,11 +126,11 @@ func Field(s string) string {
 
-func ifTerm(fmt pretty.Formatter) pretty.Formatter {
+func ifTerm(f pretty.Formatter) pretty.Formatter {
 	if !isTerm() {
 		return pretty.Nop
 	}
-	return fmt
+	return f
```
👍

johnstcn commented on cmd/cliui/main.go:
```diff
@@ -37,6 +38,51 @@ func main() {
+	}
+
+	root.Children = append(root.Children, &serpent.Command{
+		Use: "colors",
```
👍 👍

johnstcn reviewed (approved):
Nice work! Just some nits on my end.
Smoke-tested example prompts with and without colour.

DanielleMaywood committed:
chore: allow disabling color with env flags

mafredri reviewed (commented):
I think the global NoColor var could cause a race in our testing since we run things concurrently. E.g. if multiple tests in `cli` package set `--no-color` option, we'd run into it.

Other than that, I think this looks great. Might be worth looking into improving serpent so we can do this in a better way. And perhaps get rid of the package-global state in favor of a "color" instance. That's a much more intrusive change though.

DanielleMaywood commented on cli/cliui/cliui.go:
```diff
@@ -123,42 +140,45 @@ func init() {
-			pretty.FgColor(Red),
-			pretty.BgColor(color.Color("#2c2c2c")),
+			pretty.FgColor(color.Color("#ED567A")),
+			pretty.BgColor(color.Color("#2C2C2C")),
```
As we set both the foreground and background colour, we have complete control over the contrast so this looks the same regardless of the terminal theme. If we want to use an ANSI code instead we can but there isn't a need to here.

`red` was defined as `#ED567A`, and the new `red` didn't look as good on the background so I kept it as the old one.

johnstcn commented:
> I think the global NoColor var could cause a race in our testing since we run things concurrently.

@mafredri potentially, yes -- but we already check for `test.v` and skip colours. So I don't think we'd even hit this as long as the check for `test.v` came first.

DanielleMaywood commented:
> I think the global NoColor var could cause a race in our testing since we run things concurrently. E.g. if multiple tests in `cli` package set `--no-color` option, we'd run into it.

I should probably have documented this but we don't actually use the global var, I needed it to set it in `cmd/cliui` but it appears now removing it from `cmd/coder` doesn't cause the same issue so I should be able to remove this.

We can't make use of it as serpent sets it too late as we can't hook into serpent before it run commands.

mafredri commented:
> @mafredri potentially, yes -- but we already check for `test.v` and skip colours. So I don't think we'd even hit this as long as the check for `test.v` came first.

Updating the var from multiple locations is still a data race, can be seen in CI here: https://github.com/coder/coder/actions/runs/10851466684/job/30115309148?pr=14665#step:5:5121

mafredri commented:
> I should probably have documented this but we don't actually use the global var, I needed it to set it in `cmd/cliui` but it appears now removing it from `cmd/coder` doesn't cause the same issue so I should be able to remove this.

Alright, then I think using a local var to the serpent command would be better. 👍🏻

DanielleMaywood commented:
> I should probably have documented this but we don't actually use the global var, I needed it to set it in `cmd/cliui` but it appears now removing it from `cmd/coder` doesn't cause the same issue so I should be able to remove this.

Turns out serpent behaves differently with `serpent.BoolOf` versus `serpent.Discard`. BoolOf allows just writing `--no-color` but we can't do that with `Discard`.

DanielleMaywood commented:
> Alright, then I think using a local var to the serpent command would be better. 👍🏻 (+ comment explaining it's purpose)

So define a local var in `root.go` such as `noColorDiscarded`?

DanielleMaywood commented on cli/cliui/cliui.go:
```diff
@@ -37,22 +43,33 @@ var (
+		// has the flag.
+		if slices.Contains(os.Args, fmt.Sprintf("--%s", NoColorFlag)) {
+			color = termenv.Ascii
+		}
```
Unfortunately not as `NoColor` is set _after_ our code runs. This is because there is no (apparent) way to have this initialization code ran before _every_ command but after serpent sets up the options.

mafredri commented:
> So define a local var in `root.go` such as `noColorDiscarded`?

Yep, inside `func (r *RootCmd) Command(subcommands []*serpent.Command) (*serpent.Command, error)` 👍🏻.

DanielleMaywood commented on cli/cliui/cliui.go:
```diff
@@ -37,22 +43,36 @@ var (
+		// serpent as it isn't possible to create a root middleware that
+		// runs for every command. For now we just check if `os.Args`
+		// has the flag.
+		if slices.Contains(os.Args, fmt.Sprintf("--%s", NoColorFlag)) ||
```
Can do!

> `--no-color=true` and `--no-color=true`

Do my eyes deceive me or are these both the same?

mafredri replied:
A case of copy-pasta 😂, sorry. Meant `false` for the second.

DanielleMaywood replied:
Got you, just pushed a change with `--no-color=true` handling. It appears we already handle the `--no-color=false` case!

DanielleMaywood committed:
fix: apply fixes from feedback

DanielleMaywood committed:
fix: run 'make gen'

ammario commented on cli/cliui/cliui.go:
```diff
@@ -37,22 +41,37 @@ var (
 			color = termenv.Ascii
 		}
+
+		// Currently it appears there is no way to use the flag from
```
Have you tried setting the option on the root command? It should then evaluate on every command.

ammario commented on cli/cliui/cliui.go:
```diff
@@ -37,22 +41,37 @@ var (
+		// serpent as it isn't possible to create a root middleware that
+		// runs for every command. For now we just check if `os.Args`
+		// has the flag.
+		if slices.Contains(os.Args, fmt.Sprintf("--%s", NoColorFlag)) ||
```
This is super janky. Relying on os.Args makes it difficult to test. Also, someone could modify the option at the CLI layer and easily forget to update this. Have you considered checking the CLI option higher up in the stack?

DanielleMaywood replied:
I absolutely agree with you, however serpent doesn't (from what I can tell) have any way to support this as it lacks the ability to set a middleware that runs before _any_ command.

To demonstrate what I, the code in [this gist](https://gist.github.com/DanielleMaywood/6421e8aa48a160e23d5ddb84093c37c1) has a `Middleware` setup but it only runs on the root cmd and not it's children. If you know of a way to have this run before a commands handler but after options are setup I'd love to do that instead!

DanielleMaywood replied:
Scratch that: I think using `rootCmd.Walk` and injecting the middleware to every child should work? Will have a go at doing that.

DanielleMaywood replied:
I've pushed a change that should be less janky, let me know what you think!

ammario replied:
If you do the Walk approach we have the flag duplicated on every help, which is obnoxious. Can't you use an Option on the root command?

DanielleMaywood committed:
fix: refactor janky code

DanielleMaywood committed:
fix: re-add public function

DanielleMaywood committed:
fix: re-add init for non-color tests

DanielleMaywood committed:
fix: move styles to 'init' that can be

DanielleMaywood committed:
fix: stop overwriting entire DefaultStyles

DanielleMaywood committed:
fix: make code and field obey --no-color

DanielleMaywood committed:
fix: rip out '--no-color' due to race condition

We still support `NO_COLOR` env variable through termenv's
`EnvColorProfile`. The reason for the race condition is that
`DefaultStyles` is a global that we shouldn't mutate after `init`
is called, but we have to mutate it after `init` has ran to have
serpent collect the cli flags and env vars for us.

DanielleMaywood committed:
fix: apply nit

DanielleMaywood commented:
I've ripped out all the logic for `--no-color` as mutating `DefaultStyles` whilst the program is running causes a race condition that is non-trivial to solve due to using `DefaultStyles` around 150 times in the codebase. The behaviour can still be enabled by setting the `NO_COLOR` (https://no-color.org/) environment variable by using termenv's `EnvColorProfile` instead of `ColorProfile`.

ammario commented on cmd/cliui/main.go:
```diff
@@ -37,6 +38,43 @@ func main() {
 	}
 
+	root.Children = append(root.Children, &serpent.Command{
+		Use: "colors",
```
This should be hidden

DanielleMaywood replied:
Done!

ammario commented on cmd/cliui/main.go:
```diff
@@ -37,6 +38,43 @@ func main() {
+		Use: "colors",
+		Handler: func(inv *serpent.Invocation) error {
+			msg := pretty.Sprint(cliui.DefaultStyles.Code, "This is a code message")
+			_, _ = fmt.Fprintln(inv.Stdout, msg)
```
You can simplify this with `pretty.Fprintf`

DanielleMaywood replied:
Done! I had to separate the newline from the Fprintf call though as otherwise the newline would get themed.

DanielleMaywood committed:
fix: simplify code && hide command

DanielleMaywood committed:
fix: newline shouldn't be themed

DanielleMaywood committed:
fix: appease the linter

ammario reviewed (approved):
Should prob only merge when one of the core devs approve. I generally don't want to take accountability as a review since my knowledge of the internals has atrophied.

mafredri commented on cli/cliui/cliui.go:
```diff
@@ -37,17 +37,21 @@ var (
 func Color(s string) termenv.Color {
 	colorOnce.Do(func() {
-		color = termenv.NewOutput(os.Stdout).ColorProfile()
+		color = termenv.NewOutput(os.Stdout).EnvColorProfile()
```
Good solution for disabling colors 👍🏻

mafredri reviewed (approved):
Looks great. I think relying on ansi colors where we don't use a background, and hex colors where we do, is indeed the simplest approach.

Thanks for working through all the iterations of this and glad we could land on a simple solution in the end. 👍🏻
